patroni/dcs/__init__.py

```diff
@@ -85,38 +87,79 @@ def dcs_modules() -> List[str]:
         return [module_prefix + name for _, name, is_pkg in pkgutil.iter_modules([dcs_dirname]) if not is_pkg]
 
 
-def get_dcs(config: Union['Config', Dict[str, Any]]) -> 'AbstractDCS':
-    modules = dcs_modules()
+def iter_dcs_classes(
+        config: Optional[Union['Config', Dict[str, Any]]] = None
+) -> Iterator[Tuple[str, Optional[Type['AbstractDCS']]]]:
+    """Attempt to import DCS modules that are present in the given configuration.
+
+    .. note::
+            If a module successfully imports we can assume that all its requirements are installed.
+
+    :param config: configuration information with possible DCS names as keys. If given, only attempt to import DCS
+                   modules defined in the configuration. Else, if ``None``, attempt to import any supported DCS module.
 
-    for module_name in modules:
-        name = module_name.split('.')[-1]
-        if name in config:  # we will try to import only modules which have configuration section in the config file
+    :yields: a tuple containing the module ``name`` and the imported DCS class object.
+    """
+    for mod_name in dcs_modules():
+        name = mod_name.rpartition('.')[2]
+        if config is None or name in config:
             try:
-                module = importlib.import_module(module_name)
-                for key, item in module.__dict__.items():  # iterate through the module content
-                    # try to find implementation of AbstractDCS interface, class name must match with module_name
-                    if key.lower() == name and inspect.isclass(item) and issubclass(item, AbstractDCS):
-                        # propagate some parameters
-                        config[name].update({p: config[p] for p in ('namespace', 'name', 'scope', 'loop_wait',
-                                             'patronictl', 'ttl', 'retry_timeout') if p in config})
-                        # From citus section we only need "group" parameter, but will propagate everything just in case.
-                        if isinstance(config.get('citus'), dict):
-                            config[name].update(config['citus'])
-                        return item(config[name])
+                module = importlib.import_module(mod_name)
+                yield name, find_dcs_class_in_module(module)
```

Comment:
Why not yield when class is actually found?

Reply:
In fact, we only need to yield a class.
The name could be calculated using `dcs_class.__name__.lower()`

Reply:
So the iter_dcs_classes is in 2 places. One to load the module, the other to just get the available mods to generate the exception for no "suitable configuration". The second element of the tuple would need to be `None` to preserve the name of the module.

I could mix the logic up in the test for `config is None or name in config` so that either we get the name or the class object but it feels messy.

```
            module = importlib.import_module(mod_name)
            if config is None:
                yield name
            elif name in config:
                dcs_module = find_dcs_class_in_module(module)
                if dcs_module:
                    yield dcs_module
```

It feels cleaner and more consistent to always return the same object type and tuple structure.